File: abmt2025/src/main/java/abmt2025/project/utils/AnalyzePtStopsInDrtArea.java
package abmt2025.project.utils;

import org.geotools.data.simple.SimpleFeatureCollection;
import org.geotools.data.simple.SimpleFeatureSource;
import org.geotools.feature.FeatureIterator;
import org.locationtech.jts.geom.Geometry;
import org.locationtech.jts.geom.Point;
import org.matsim.api.core.v01.Coord;
import org.matsim.api.core.v01.Scenario;
import org.matsim.core.config.Config;
import org.matsim.core.config.ConfigUtils;
import org.matsim.core.scenario.ScenarioUtils;
import org.matsim.core.utils.geometry.geotools.MGC;
import org.matsim.core.utils.gis.ShapeFileReader;
import org.matsim.pt.transitSchedule.api.TransitSchedule;
import org.matsim.pt.transitSchedule.api.TransitStopFacility;
import org.opengis.feature.simple.SimpleFeature;

import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Utility class per analizzare quali fermate PT sono dentro o vicino alla zona DRT.
 *
 * Uso:
 *   java -cp target/abmt2025.jar abmt2025.project.utils.AnalyzePtStopsInDrtArea \
 *       /path/to/config.xml \
 *       /path/to/drt_shapefile.shp \
 *       5000 \
 *       /path/to/output.csv
 */
public class AnalyzePtStopsInDrtArea {

    public static void main(String[] args) {
        if (args.length < 2) {
            System.out.println("Uso: AnalyzePtStopsInDrtArea <config.xml> <drt_shapefile.shp> [radius_m] [output.csv]");
            System.out.println();
            System.out.println("Argomenti:");
            System.out.println("  config.xml      - MATSim config file con transit schedule");
            System.out.println("  drt_shapefile   - Shapefile della zona DRT");
            System.out.println("  radius_m        - Raggio di ricerca intermodale (default: 5000)");
            System.out.println("  output.csv      - File CSV di output (opzionale)");
            return;
        }

        String configPath = args[0];
        String shapefilePath = args[1];
        double radiusM = args.length > 2 ? Double.parseDouble(args[2]) : 5000.0;
        String outputPath = args.length > 3 ? args[3] : null;

        // Carica config e scenario
        System.out.println("Caricamento config: " + configPath);
        Config config = ConfigUtils.loadConfig(configPath);
        Scenario scenario = ScenarioUtils.loadScenario(config);
        TransitSchedule schedule = scenario.getTransitSchedule();

        System.out.println("Fermate PT nel transit schedule: " + schedule.getFacilities().size());

        // Carica shapefile
        System.out.println("Caricamento shapefile DRT: " + shapefilePath);
        Geometry drtArea = loadShapefile(shapefilePath);
        if (drtArea == null) {
            System.err.println("Errore nel caricamento dello shapefile");
            return;
        }

        Geometry drtAreaBuffered = drtArea.buffer(radiusM);

        // Analizza le fermate
        List<StopAnalysis> results = new ArrayList<>();
        int insideCount = 0;
        int withinRadiusCount = 0;

        for (TransitStopFacility stop : schedule.getFacilities().values()) {
            Coord coord = stop.getCoord();
            Point point = MGC.coord2Point(coord);

            boolean insideDrt = drtArea.contains(point);
            boolean withinRadius = drtAreaBuffered.contains(point);
            double distance = insideDrt ? 0 : point.distance(drtArea);

            String stopCategory = (String) stop.getAttributes().getAttribute("stopCategory");

            StopAnalysis analysis = new StopAnalysis(
                stop.getId().toString(),
                stop.getName(),
                coord.getX(),
                coord.getY(),
                stopCategory,
                insideDrt,
                withinRadius,
                distance
            );
            results.add(analysis);

            if (insideDrt) insideCount++;
            if (withinRadius) withinRadiusCount++;
        }

        // Ordina per distanza
        results.sort(Comparator.comparingDouble(a -> a.distanceToArea));

        // Stampa riepilogo
        System.out.println();
        System.out.println("================================================================");
        System.out.println("RIEPILOGO FERMATE PT vs ZONA DRT");
        System.out.println("================================================================");
        System.out.println("Totale fermate PT: " + results.size());
        System.out.println("Fermate DENTRO la zona DRT: " + insideCount);
        System.out.println("Fermate entro " + (int)radiusM + "m dalla zona DRT: " + withinRadiusCount);
        System.out.println("Fermate oltre " + (int)radiusM + "m dalla zona DRT: " + (results.size() - withinRadiusCount));
        System.out.println("================================================================");

        if (insideCount == 0) {
            System.out.println();
            System.out.println("⚠️  ATTENZIONE: Nessuna fermata PT dentro la zona DRT!");
            System.out.println("   I viaggi multi-modali usano fermate FUORI dalla zona.");
        }

        // Mostra le fermate più vicine
        System.out.println();
        System.out.println("Fermate PT più vicine alla/dentro la zona DRT:");
        int shown = 0;
        for (StopAnalysis stop : results) {
            if (shown >= 15) break;
            String status = stop.insideArea ? "DENTRO" : String.format("%.0fm", stop.distanceToArea);
            String cat = stop.category != null ? " (cat:" + stop.category + ")" : "";
            System.out.println("  - " + stop.name + " [" + stop.id + "]: " + status + cat);
            shown++;
        }

        // Salva CSV se richiesto
        if (outputPath != null) {
            saveToCSV(results, outputPath);
            System.out.println();
            System.out.println("Risultati salvati in: " + outputPath);
        }
    }

    private static Geometry loadShapefile(String shapefilePath) {
        Geometry mergedGeometry = null;

        try {
            SimpleFeatureSource featureSource = ShapeFileReader.readDataFile(shapefilePath);
            SimpleFeatureCollection featureCollection = featureSource.getFeatures();

            try (FeatureIterator<SimpleFeature> features = featureCollection.features()) {
                while (features.hasNext()) {
                    SimpleFeature feature = features.next();
                    Geometry geometry = (Geometry) feature.getDefaultGeometry();
                    if (geometry != null) {
                        if (mergedGeometry == null) {
                            mergedGeometry = geometry;
                        } else {
                            mergedGeometry = mergedGeometry.union(geometry);
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }

        return mergedGeometry;
    }

    private static void saveToCSV(List<StopAnalysis> results, String outputPath) {
        try (PrintWriter writer = new PrintWriter(new FileWriter(outputPath))) {
            writer.println("stop_id,name,x,y,category,inside_drt_area,within_intermodal_radius,distance_to_drt_m");
            for (StopAnalysis stop : results) {
                writer.printf("%s,%s,%.1f,%.1f,%s,%b,%b,%.1f%n",
                    stop.id,
                    stop.name.replace(",", ";"),
                    stop.x, stop.y,
                    stop.category != null ? stop.category : "",
                    stop.insideArea,
                    stop.withinRadius,
                    stop.distanceToArea);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static class StopAnalysis {
        String id;
        String name;
        double x, y;
        String category;
        boolean insideArea;
        boolean withinRadius;
        double distanceToArea;

        StopAnalysis(String id, String name, double x, double y, String category,
                     boolean insideArea, boolean withinRadius, double distanceToArea) {
            this.id = id;
            this.name = name;
            this.x = x;
            this.y = y;
            this.category = category;
            this.insideArea = insideArea;
            this.withinRadius = withinRadius;
            this.distanceToArea = distanceToArea;
        }
    }
}

File: abmt2025/scripts/analyze_pt_stops_in_drt_area.py
#!/usr/bin/env python3
"""
Script per verificare le fermate PT rispetto alla zona DRT.

Questo script:
1. Carica lo shapefile della zona DRT
2. Carica il transit_schedule.xml.gz
3. Verifica quali fermate PT sono:
   - DENTRO la zona DRT
   - Entro un certo raggio dalla zona DRT (es. 5km come da config intermodale)
4. Genera un report

USO:
    python analyze_pt_stops_in_drt_area.py \
        --schedule /path/to/transit_schedule.xml.gz \
        --shapefile /path/to/drt_zone.shp \
        --radius 5000 \
        --output /path/to/output_report.csv
"""

import argparse
import gzip
import xml.etree.ElementTree as ET
from pathlib import Path

try:
    import geopandas as gpd
    from shapely.geometry import Point
    from shapely.ops import unary_union
    HAS_GEO = True
except ImportError:
    HAS_GEO = False
    print("ATTENZIONE: geopandas/shapely non installati.")
    print("Installa con: pip install geopandas shapely")


def parse_transit_schedule(schedule_path):
    """Parse transit_schedule.xml.gz e estrae le fermate."""
    stops = []

    # Apri file compresso o normale
    if schedule_path.endswith('.gz'):
        f = gzip.open(schedule_path, 'rt', encoding='utf-8')
    else:
        f = open(schedule_path, 'r', encoding='utf-8')

    try:
        tree = ET.parse(f)
        root = tree.getroot()

        # Cerca stopFacility elementi
        for stop in root.iter('stopFacility'):
            stop_id = stop.get('id')
            x = float(stop.get('x'))
            y = float(stop.get('y'))
            name = stop.get('name', 'N/A')

            # Cerca attributi aggiuntivi (es. stopCategory)
            stop_category = None
            for attr in stop.iter('attribute'):
                if attr.get('name') == 'stopCategory':
                    stop_category = attr.text
                    break

            stops.append({
                'id': stop_id,
                'x': x,
                'y': y,
                'name': name,
                'stopCategory': stop_category
            })
    finally:
        f.close()

    return stops


def analyze_stops_vs_drt_area(stops, shapefile_path, radius_m=5000):
    """Analizza le fermate rispetto alla zona DRT."""

    if not HAS_GEO:
        print("Impossibile eseguire analisi spaziale senza geopandas")
        return None

    # Carica shapefile
    gdf_drt = gpd.read_file(shapefile_path)
    drt_area = unary_union(gdf_drt.geometry)

    # Buffer per il raggio di ricerca intermodale
    drt_area_buffered = drt_area.buffer(radius_m)

    results = []
    for stop in stops:
        point = Point(stop['x'], stop['y'])

        inside_drt = drt_area.contains(point)
        inside_buffer = drt_area_buffered.contains(point)
        distance_to_drt = point.distance(drt_area) if not inside_drt else 0

        results.append({
            **stop,
            'inside_drt_area': inside_drt,
            'within_intermodal_radius': inside_buffer,
            'distance_to_drt_area_m': round(distance_to_drt, 1)
        })

    return results


def print_summary(results, radius_m):
    """Stampa un riepilogo dei risultati."""

    total = len(results)
    inside = sum(1 for r in results if r['inside_drt_area'])
    within_radius = sum(1 for r in results if r['within_intermodal_radius'])
    outside = total - within_radius

    print("\n" + "="*60)
    print("RIEPILOGO FERMATE PT vs ZONA DRT")
    print("="*60)
    print(f"Totale fermate PT nel transit_schedule: {total}")
    print(f"Fermate DENTRO la zona DRT: {inside}")
    print(f"Fermate entro {radius_m}m dalla zona DRT: {within_radius}")
    print(f"Fermate oltre {radius_m}m dalla zona DRT: {outside}")
    print("="*60)

    if inside == 0:
        print("\n⚠️  ATTENZIONE: Nessuna fermata PT dentro la zona DRT!")
        print("   I viaggi multi-modali usano fermate FUORI dalla zona.")

    # Mostra le fermate più vicine fuori dalla zona
    outside_stops = [r for r in results if not r['inside_drt_area'] and r['within_intermodal_radius']]
    if outside_stops:
        print(f"\nFermate PT raggiungibili con DRT (entro {radius_m}m):")
        outside_stops.sort(key=lambda x: x['distance_to_drt_area_m'])
        for stop in outside_stops[:10]:
            print(f"  - {stop['name']} ({stop['id']}): {stop['distance_to_drt_area_m']:.0f}m dalla zona")

    # Mostra le fermate dentro la zona
    inside_stops = [r for r in results if r['inside_drt_area']]
    if inside_stops:
        print(f"\nFermate PT DENTRO la zona DRT:")
        for stop in inside_stops:
            cat = f" (cat: {stop['stopCategory']})" if stop['stopCategory'] else ""
            print(f"  - {stop['name']} ({stop['id']}){cat}")


def save_to_csv(results, output_path):
    """Salva i risultati in CSV."""
    import csv

    with open(output_path, 'w', newline='', encoding='utf-8') as f:
        if results:
            writer = csv.DictWriter(f, fieldnames=results[0].keys())
            writer.writeheader()
            writer.writerows(results)

    print(f"\nRisultati salvati in: {output_path}")


def main():
    parser = argparse.ArgumentParser(description='Analizza fermate PT vs zona DRT')
    parser.add_argument('--schedule', required=True, help='Path al transit_schedule.xml.gz')
    parser.add_argument('--shapefile', required=True, help='Path allo shapefile della zona DRT')
    parser.add_argument('--radius', type=int, default=5000,
                        help='Raggio di ricerca intermodale in metri (default: 5000)')
    parser.add_argument('--output', help='Path per il file CSV di output')

    args = parser.parse_args()

    print(f"Caricamento transit_schedule da: {args.schedule}")
    stops = parse_transit_schedule(args.schedule)
    print(f"Trovate {len(stops)} fermate PT")

    if HAS_GEO:
        print(f"\nAnalisi spaziale con shapefile: {args.shapefile}")
        results = analyze_stops_vs_drt_area(stops, args.shapefile, args.radius)

        if results:
            print_summary(results, args.radius)

            if args.output:
                save_to_csv(results, args.output)
    else:
        print("\nSolo lista fermate (senza analisi spaziale):")
        for stop in stops[:20]:
            print(f"  {stop['id']}: {stop['name']} ({stop['x']}, {stop['y']})")
        if len(stops) > 20:
            print(f"  ... e altre {len(stops) - 20} fermate")


if __name__ == '__main__':
    main()
